indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -629,6 +648,207 @@ public Response getCompleteTasks(
+          );
+        case "COMPLETE":
+          final List<TaskStatusPlus> completedTasks = getCompletedTasks(interval, maxCompletedTasks, dataSource, req);
+          final List<TaskStatusPlus> authorizedList = securedTaskStatusPlus(
+              completedTasks,
+              dataSource,
+              req
+          );
+          return Response.ok(authorizedList).build();
+        default:
+          return Response.status(Status.BAD_REQUEST)
+                         .entity("Bad state query param passed : " + state).entity(ImmutableList.of()).build();
+      }
+    }
+    // if no state passed, get all tasks
+    final List<TaskRunnerWorkItem> allActiveTasks = getAllActiveTasks(req);
+    final List<TaskStatusPlus> completedTasks = getCompletedTasks(interval, maxCompletedTasks, dataSource, req);
+    List activeTasks = new ArrayList(allActiveTasks);
+    Function<TaskRunnerWorkItem, TaskStatusPlus> transformFunc = workItem -> new TaskStatusPlus(
+        workItem.getTaskId(),
+        workItem.getTaskType(),
+        workItem.getCreatedTime(),
+        workItem.getQueueInsertionTime(),
+        workItem.getTaskState(),
+        null,
+        workItem.getLocation(),
+        workItem.getDataSource(),
+        null
+    );
+    List<TaskStatusPlus> transformedActiveList = Lists.transform(activeTasks, transformFunc);
+    final List<TaskStatusPlus> allTasks = Lists.newArrayList();
+    allTasks.addAll(transformedActiveList);
+    allTasks.addAll(completedTasks);
+
+    Collection<? extends TaskStatusPlus> authorizedList = securedTaskStatusPlus(
+        allTasks,
+        dataSource,
+        req
+    );
+    return Response.ok(authorizedList).build();
+  }
+
+  private List<TaskStatusPlus> getCompletedTasks(
+      String interval,
+      Integer maxCompletedTasks,
+      String dataSource,
+      HttpServletRequest req
+  )
+  {
+    Duration duration = null;
+    if (interval != null) {
+      final Interval theInterval = Intervals.of(interval.replace("_", "/"));
+      duration = theInterval.toDuration();
+    }
+
+    final Function<String, Task> taskFunction = taskId -> {
+      final Optional<Task> optionalTask = taskStorageQueryAdapter.getTask(taskId);
+      if (!optionalTask.isPresent()) {
+        throw new WebApplicationException(
+            Response.serverError().entity(
+                StringUtils.format("No task information found for task with id: [%s]", taskId)
+            ).build()
+        );
+      }
+      return optionalTask.get();
+    };
+
+    final List<TaskStatus> recentlyFinishedTasks = taskStorageQueryAdapter.getRecentlyFinishedTaskStatuses(
+            maxCompletedTasks, duration);
+
+    final List<TaskStatusPlus> completeTasks = Lists.newArrayList(Iterables.transform(
+        recentlyFinishedTasks,
+        status -> {
+          final Pair<DateTime, String> pair = taskStorageQueryAdapter.getCreatedDateAndDataSource(status.getId());
```

Comment:
This is still doing `taskStorageQueryAdapter.getCreatedDateAndDataSource` for every single complete task. Let's try to structure the code to avoid this somehow. Possibly by adding a method to `taskStorageQueryAdapter` that returns this info for all complete tasks at once.